server/lib/chat/chat-service.ts

```diff
@@ -176,25 +201,34 @@ export default class ChatService {
     let channel: ChannelInfo | undefined
     let userChannelEntry: UserChannelEntry | undefined
     let message: ChatMessage
+    let error: ChatServiceError | undefined
     do {
       attempts += 1
       try {
         await transact(async client => {
           channel = await findChannelByName(channelName, client)
           if (channel) {
             if (this.state.users.has(userId) && this.state.users.get(userId)!.has(channel.id)) {
-              throw new ChatServiceError(
+              error = new ChatServiceError(
```

Comment:
This is an incorrect change, you *want* to throw this here, you *want* this transaction to be rolled back. The fix was to pass back a flag (either by setting something from the closure, or by returning a value (`transact` will return whatever you return from the transaction function), and checking that to see if the user was banned. It was *not* to avoid throwing *every* error

Reply:
So thinking about this further, I think handling things in this way is really a mistake to begin with. Generally you should try to do things idempotently: 2 requests for the same thing coming in should deliver the same result. If I as a user request to join a channel twice, and the 2nd request hits at a point where I've already joined that channel, why would I care that I'm already in the channel? Better for the server to tell me "sure, you're in that channel". Programming it to be an error makes it very easy to do bad things on the client (e.g. show the user an error dialog instead of just... taking them to the channel they wanted to be in).

So, don't write things this way, this is not an error.

Your "channel exists" code here should have 3 cases:
1) The user is already in the channel: cool, we're done, mark this as successful and return
2) The user is banned or becomes banned: cool, we're done, mark that they are banned and return (throw an error corresponding to being banned outside of the transaction)
3) The user is not in the channel: do the things necessary to have them join, return

Reply:
I'll also note that I don't really think checking `this.state` to determine if the user is in the channel is a great idea here. That state is not in sync with what is in the DB, it is only updated *after* DB operations complete, with some non-determinant amount of time between their completion and the update. Would almost certainly be better to actually query the DB for this information here.

(Aside from the out-of-sync stuff, I think `this.state.users` only contains users that currently have a connected websocket? So it's even worse to use that for this purpose)